popout.js

```diff
@@ -1,241 +1,355 @@
+        this.ID = [...Array(24)].map(() => Math.floor(Math.random() * 16).toString(16)).join("");
+    }
+
+    log(msg, ...args) {
+        if (game && game.settings.get("popout", "verboseLogs")) {
+            const color = "background: #6699ff; color: #000; font-size: larger;";
+            console.debug(`%c PopoutModule: ${msg}`, color, ...args);
+        }
+    }
+
+    async init() {
+        game.settings.register("popout", "useWindows", {
+            name: "Pop sheets out into windows",
+            hint: "Force the popped out sheet to be a window with minimal decorations. Otherwise uses your browser's default (a new tab most likely)",
+            scope: "client",
+            config: true,
+            default: false,
+            type: Boolean
+        });
+
+        game.settings.register("popout", "verboseLogs", {
+            name: "Enable more module logging.",
+            hint: "Enables more verbose module logging. This is useful for debugging the module. But otherwise should be left off.",
+            scope: "client",
+            config: true,
+            default: false,
+            type: Boolean
+        });
+
+
+        // We replace the games window registry with a proxy object so we can intercept
+        // every new application window creation event.
+        const handler = {
+            set: (obj, prop, value) => {
+                const result = Reflect.set(obj, prop, value);
+                this.log("Intercept ui-window create", value)
+                try {
+                    if (value && value.options && value.options.popOut) {
+                        this.addPopout(value);
+                    }
+                } catch (err) {
+                    // We must never fail here.
+                    this.log(err);
+                }
+                return result
+            }
+        };
+        ui.windows = new Proxy(ui.windows, handler);
+        this.log("Installed window interceptor", ui.windows);
+
+        // HACK(aposney: 2020-07-12): we need to init tinymce to ensure it's plugins 
+        // are loaded into the frame. Otherwise our popouts will not be able to access
+        // the lazy loaded javascript mce plugins. 
+        const elem = $("<div style=\"display: none;\"><p id=\"mce_init\"> foo </p></div>")
+        $("body").append(elem)
+        const config = { target: elem[0], plugins: CONFIG.TinyMCE.plugins };
+        const editor = await tinyMCE.init(config);
+        editor[0].remove();
+    }
+
+    addPopout(app) {
+        this._addPopout(app, 0);
+    }
+
+
+    _addPopout(app, recurse) {
+        if (recurse > this.MAX_TIMEOUT) {
+            this.log("Timeout out waiting for app to render");
+            return;
+        }
+
+        if (!(app._state == Application.RENDER_STATES.RENDERED)) {
+            window.setTimeout(() => {
+                this._addPopout(app, recurse + this.TIMEOUT_INTERVAL);
+            }, this.TIMEOUT_INTERVAL);
+            return;
+        }

```

Comment:
I originally used the recursive method because I didn't think you could use aync functions in JS proxies, but you are right, this solution is clearer.